File: vetix/plugins/reverse_shell.py
import re

from vetix.plugin import Issue, Plugin, Severity
from vetix.utils.utils import get_relative_path


class ReverseShellPlugin(Plugin):
    """Detect reverse shell patterns (bash /dev/tcp, netcat -e, socat exec:)."""

    name = "Reverse Shell"

    _SHELL_RE = re.compile(
        r"(?i)(/dev/(tcp|udp)/|\bnc\s+.*\s-e\s+|\bsocat\b.*\bexec:)"
    )

    def scan(self, skill_dir: str, file_path: str, content: str) -> list[Issue]:
        issues: list[Issue] = []
        relative_path = get_relative_path(file_path, skill_dir)

        for match in self._SHELL_RE.finditer(content):
            line = content.count("\n", 0, match.start()) + 1
            issues.append(Issue(
                name="Reverse shell",
                severity=Severity.CRITICAL,
                category="Remote Execution",
                description=(
                    f"Reverse shell pattern detected ({match.group(0)})."
                ),
                file_path=relative_path,
                line=line,
                audit_required=True,
            ))
        return issues

File: internal/plugin/reverse_shell.go
package plugin

import (
	"fmt"
	"regexp"
)

var reverseShellRe = regexp.MustCompile(`(?i)(/dev/(tcp|udp)/|\bnc\s+.*\s-e\s+|\bsocat\b.*\bexec:)`)

type ReverseShellPlugin struct{}

func (ReverseShellPlugin) Meta() Meta {
	return Meta{
		ID:          "reverse_shell",
		Name:        "Reverse Shell",
		Description: "Detects reverse-shell patterns such as /dev/tcp redirections, nc -e and socat exec:.",
	}
}

func (ReverseShellPlugin) Scan(skillDir, filePath, content string) []Issue {
	var issues []Issue
	rel := relativePath(filePath, skillDir)
	for _, m := range reverseShellRe.FindAllStringIndex(content, -1) {
		issues = append(issues, Issue{
			Name:          "Reverse shell",
			Severity:      SeverityCritical,
			Category:      CatRemoteExecution,
			Description:   fmt.Sprintf("Reverse shell pattern detected (%s).", content[m[0]:m[1]]),
			FilePath:      rel,
			Line:          lineOf(content, m[0]),
			AuditRequired: true,
		})
	}
	return issues
}
